Add support for Enclustra Mercury XU* modules

kgugala opened the pull request:
This PR adds suport for the Cortex R5 CPU in the Xilinx Zynq MP SoC on Enclustra's Mercury XU* Modules.

It also adds possibility to disable ECC checking in the Cortex R Tightly Coupled Memory (TCM)

kgugala commented:
ping @galak @ioannisg

kgugala commented:
This PR is hanging here for quite some time and it is blocking our next PRs adding AMP support for Cortexes-R. I have rebased the PR onto current master. @galak @ioannisg can you take a look on this one, or point somebody else who can review this code? Thanks in advance

ioannisg commented on arch/arm/core/cortex_r/Kconfig:
```diff
@@ -86,6 +86,12 @@ config GEN_ISR_TABLES
 
+config DISABLE_TCM_ECC
+	bool
+	default n
```
This option is introduced here, right?
Any reason this is hidden?
Also, default n is not required for boolean Kconfigs

ioannisg commented on arch/arm/core/cortex_r/reset.S:
```diff
@@ -99,4 +100,8 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
 #endif
 
+#if defined(CONFIG_DISABLE_TCM_ECC)
+	bl disable_ecc
```
Is this something that is SoC specific or ARCH specific?
If SoC specific, I would suggesting doing this under PLATFORM_SPECIFIC_INIT (see for example Cortex-M reset.S)

stephanosio replied:
This should be at arch-level. TCM ECC is a feature that is directly integrated into the Cortex-R processors.

ioannisg commented on boards/arm/mercury_xu/mercury_xu_board_defconfig:
```diff
@@ -0,0 +1,19 @@
+CONFIG_BOARD_MERCURY_XU_BOARD=y
+
+CONFIG_ISR_STACK_SIZE=1024
+CONFIG_THREAD_STACK_INFO=y
```
Why was this needed here?

stephanosio commented:
This PR looks ancient. Is this still being worked on?

stephanosio commented on arch/arm/core/cortex_r/tcm.c:
```diff
@@ -0,0 +1,14 @@
+
+void disable_ecc(void)
+{
+	__asm__(
+		"mrc p15, 0, r0, c1, c0, 1\n\t" /* read ACTLR */
+		"bic r0, r0, #0xe000000\n\t"  /* disable ecc */
+		"mcr p15, 0, r0, c1, c0, 1\n\t" /* write ACTLR */
+	);
```
This can make use of the `__{get,set}_ACTLR` functions.

https://github.com/zephyrproject-rtos/cmsis/blob/a2b82e79bd8ac156fb283d5f52fa0a2373962cf9/CMSIS/Core_R/Include/cmsis_cp15.h#L35-L51

stephanosio reviewed (changes requested):
Needs rebase

stephanosio commented on boards/arm/mercury_xu/Kconfig.board:
```diff
@@ -0,0 +1,6 @@
+config BOARD_MERCURY_XU_BOARD
+	bool "Mercury XU Board"
+	depends on SOC_XILINX_ZYNQMP
+	select DISABLE_TCM_ECC
```
Unless there is a board-level reason as to why `DISABLE_TCM_ECC` must be disabled, it should be specified in the `defconfig` (and yes, the symbol should have a prompt as @ioannisg pointed out).

stephanosio commented on boards/arm/mercury_xu/board.h:
```diff
@@ -0,0 +1,10 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef __INC_BOARD_H
```
```suggestion
#ifndef ZEPHYR_BOARDS_ARM_MERCURY_XU_BOARD_H_
```

stephanosio commented on boards/arm/mercury_xu/mercury_xu_board.yaml:
```diff
@@ -0,0 +1,18 @@
+arch: arm
+toolchain:
+  - zephyr
+  - gccarmemb
```
`gccarmemb` is obsolete.

```suggestion
  - gnuarmemb
```

stephanosio commented on boards/arm/mercury_xu/mercury_xu_board.yaml:
```diff
@@ -0,0 +1,18 @@
+  - gccarmemb
+ram: 64
+testing:
+  ignore_tags:
+   - nmi
+   - benchmark
+   - interrupt
+   - bluetooth
+   - socket
+   - crypto
+   - net
```
These are probably no longer necessary.

kgugala commented:
@stephanosio @ioannisg  thanks for the comments. We'll rebase this and test again

kgugala commented:
@stephanosio @ioannisg I have rebased the PR. I believe the comments are addressed. Can you take a look on this one?

stephanosio commented on arch/arm/include/aarch32/cortex_r/tcm.h:
```diff
@@ -0,0 +1,45 @@
+/*
```
Please add this file to `arch/arm/include/aarch32/cortex_a_r` directory (`cortex_r` directory will be removed in future releases).

kgugala replied:
done

stephanosio commented on arch/arm/core/aarch32/cortex_a_r/tcm.c:
```diff
@@ -0,0 +1,16 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <cmsis_cp15.h>
```
Please include `arch/arm/aarch32/cortex_a_r/cmsis.h`

kgugala replied:
done

stephanosio commented on arch/arm/core/aarch32/cortex_a_r/tcm.c:
```diff
@@ -0,0 +1,16 @@
+	uint32_t actlr;
+
+	actlr = __get_ACTLR();
+	actlr &= ~0xe000000;
```
Given that these bits are defined in the CMSIS, please make use of the definitions:
https://github.com/zephyrproject-rtos/cmsis/blob/a2b82e79bd8ac156fb283d5f52fa0a2373962cf9/CMSIS/Core_R/Include/core_cr.h#L511-L518

```suggestion
	actlr &= ~(ACTLR_ATCMPCEN_Msk | ACTLR_B0TCMPCEN_Msk | ACTLR_B1TCMPCEN_Msk);
```

kgugala replied:
done

kgugala commented:
I have rebased the PR on top of current master. @stephanosio can you check if your comments have been resolved?

PiotrZierhoffer commented on arch/arm/include/aarch32/cortex_a_r/tcm.h:
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2019, Antmicro
```
I believe these should all be 2020 now

kgugala replied:
should be fixed now (here and in all the other occurrences)

PiotrZierhoffer commented on boards/arm/mercury_xu/Kconfig.defconfig:
```diff
@@ -0,0 +1,19 @@
+config BOARD
+	default "mercury_xu"
+
+config BUILD_OUTPUT_STRIPPED
```
Is this needed? I see this setting in 3 x86 boards, but I don't think it should be a default

kgugala replied:
indeed, this is not needed. Removed

PiotrZierhoffer reviewed (approved):
LGTM, minor comments only

kgugala commented:
@stephanosio can you re-review this PR? I believe all your comments were addressed

stephanosio reviewed (commented):
Looks ok overall.

wsipak committed:
boards: arm: add suport for Enclustra Mercury XU boards

This commit adds support for Enclustra's boards with ZynqMP SoC

Signed-off-by: Wojciech Sipak <wsipak@antmicro.com>

wsipak committed:
arch: arm: cortex_r: disable ECC on TCMs

This commit adds possibility to disable ECC in Tightly Coupled
Memory in Cortex-R.
Linker scripts places stacks in this memory and marks it as
.noinit section. With ECC enabled, stack read accesses without
previous write result in Data Abort Exception.

Signed-off-by: Wojciech Sipak <wsipak@antmicro.com>

wsipak commented:
@stephanosio Thank you. Your suggested changes are included in the code now.

wsipak commented on arch/arm/include/aarch32/cortex_a_r/tcm.h:
```diff
@@ -0,0 +1,45 @@
+
+#endif /* _ASMLANGUAGE */
+
+#endif /* ZEPHYR_ARCH_ARM_INCLUDE_AARCH32_CORTEX_R_TCM_H_ */
```
Thank you. This should be ok now.

stephanosio reviewed (approved):
Looks ok to me.

p.s. it would be nice to have docs.
